### ISSUE TITLE
[metal] (performance) Late MTLBuffer creation

### ISSUE THREAD

**Issue by nical** [gfx-rs/wgpu#3082]
**Description**

Mac's integrated gpu tooling reports that "Your application created a MTLBuffer object during GPU work (wgpu internal) Staging" and suggests creating state objects at load time for best performance.

It's not clear how much this impacts performance in practice, just making note of it for future reference.

**Platform**

MacOs, wgpu 0.14

**Comment by kpreid** [comment]
I believe that the _advice_ in this message can be considered to mean essentially “Hey, how about if you loaded all your data before the game started, instead of after?”, which is not particularly solvable by wgpu itself.

It's possible to reduce creation of buffers by replacing `write_buffer()` with [`wgpu::util::StagingBelt`](https://wgpu.rs/doc/wgpu/util/struct.StagingBelt.html), which reuses staging buffers (but still creates new ones if needed to fit the data supplied) as well as combining multiple small writes into fewer buffers — and I've found for myself that this does in fact improve performance when writing lots of buffer data per frame, with the Metal backend. However, `StagingBelt` currently requires tuning of the staging buffer size (if not, it can substantially over-allocate, #1441), and it doesn't yet support writing textures because of their complex stride requirements (#1444).

If I remember correctly, @cwfitzgerald has talked about someday making `write_buffer()` incorporate the ideas of `StagingBelt` in some way.

**Comment by cwfitzgerald** [comment]
I agree this isn't particularly actionable right now - once we implement write_buffer staging buffer reuse it'll get a bit better, but this is more (wishful) advice to the end user.

**Comment by nical** [comment]
> I believe that the advice in this message can be considered to mean essentially “Hey, how about if you loaded all your data before the game started, instead of after?”, which is not particularly solvable by wgpu itself.

I believe it really means "try not to allocate buffer and other resources in the critical path", regardless of when you use them to get data across.

**Comment by jimblandy** [comment]
Treating this as wgpu issue, not a user issue, because wgpu does decide when to create staging buffers.
